interfaces/builtin/pulseaudio.go

```diff
@@ -20,30 +20,164 @@
 /etc/pulse/ r,
 /etc/pulse/* r,
-/{run,dev}/shm/pulse-shm-* rwk,
-owner /{,var/}run/user/*/pulse/ r,
-owner /{,var/}run/user/*/pulse/native rwk,
-`
+owner @{HOME}/.pulse-cookie rk,
+owner @{HOME}/.config/pulse/cookie rk,
+`)
 
-const pulseaudioConnectedPlugSecComp = `
+var pulseaudioConnectedPlugSecComp = []byte(`
 setsockopt
+getsockopt
 connect
 sendto
-`
-
-// NewPulseAudioInterface returns a new "pulseaudio" interface.
-func NewPulseAudioInterface() interfaces.Interface {
-	return &commonInterface{
-		name: "pulseaudio",
-		connectedPlugAppArmor: pulseaudioConnectedPlugAppArmor,
-		connectedPlugSecComp:  pulseaudioConnectedPlugSecComp,
-		reservedForOS:         true,
-		autoConnect:           true,
+shmctl
+getsockname
+getpeername
+sendmsg
+recvmsg
+`)
+
+var pulseaudioPermanentSlotAppArmor = []byte(`
+# When running PulseAudio in system mode it will switch to the at
+# build time configured user/group on startup.
+capability setuid,
+capability setgid,
+
+capability sys_nice,
+capability sys_resource,
+
+owner @{PROC}/@{pid}/exe r,
+/etc/machine-id r,
+
+# Audio related
+@{PROC}/asound/devices r,
+@{PROC}/asound/card** r,
+/dev/snd/pcm* rw,
+/dev/snd/control* rw,
+/dev/snd/timer r,
+/sys/**/sound/** r,
+
+# For udev
+network netlink raw,
+/sys/devices/virtual/dmi/id/sys_vendor r,
+/sys/devices/virtual/dmi/id/bios_vendor r,
+# FIXME: use udev queries to make this more specific
+/run/udev/data/** r,
+
+owner /{,var/}run/pulse/ rw,
+owner /{,var/}run/pulse/** rwk,
+
+# Shared memory based communication with clients
+/{run,dev}/shm/pulse-shm-* rwk,
+`)
+
+var pulseaudioPermanentSlotSecComp = []byte(`
+# The following are needed for UNIX sockets
+personality
+setpriority
+setsockopt
+getsockname
+bind
+listen
+sendto
+recvfrom
+accept4
+shmctl
+getsockname
+getpeername
+sendmsg
+recvmsg
+# Needed to set root as group for different state dirs
+# pulseaudio creates on startup.
+setgroups
+`)
+
+type PulseAudioInterface struct{}
+
+func (iface *PulseAudioInterface) Name() string {
+	return "pulseaudio"
+}
+
+func (iface *PulseAudioInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *PulseAudioInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		// If we're running on classic then allow access to the pulse config
+		// directory
+		if release.OnClassic {
+			b := bytes.NewBuffer(pulseaudioConnectedPlugAppArmorDesktop)
+			// Add these bytes to the connected plug apparmor rules
+			b.Write([]byte("owner /{,var/}run/user/*/pulse/ rwk,\n"))
+			b.Write([]byte("owner /{,var/}run/user/*/pulse/native rwk,\n"))
```

Comment:
The difference with other interfaces is that they are different rules (if not, then that was a mistake). This is very clearly "always use these few rules and if on classic, also add these others", so the code should reflect that by putting the "always use these few rules" in pulseaudioConnectedPlugAppArmor and then the desktop only ones in pulseaudioConnectedPlugAppArmorDesktop. As it stands there is a small maintenance and auditibility issue since the `/{,var/}run/user/*/pulse/` rules are duplicated in two quite different places.